src/node_crypto.cc

```diff
@@ -1323,6 +1331,188 @@ void SecureContext::GetCertificate(const FunctionCallbackInfo<Value>& args) {
+    const FunctionCallbackInfo<Value>& args) {
+  SecureContext* wrap = Unwrap<SecureContext>(args.Holder());
+
+  SSL_CTX_set_psk_server_callback(wrap->ctx_,
+                                  SecureContext::PskServerCallback);
+  SSL_CTX_set_psk_client_callback(wrap->ctx_,
+                                  SecureContext::PskClientCallback);
+}
+
+unsigned int SecureContext::PskServerCallback(SSL *ssl,
+                                              const char *identity,
+                                              unsigned char *psk,
+                                              unsigned int max_psk_len) {
+  SecureContext* sc = static_cast<SecureContext*>(
+      SSL_CTX_get_app_data(ssl->ctx));
+
+  Environment* env = sc->env();
+  Isolate* isolate = env->isolate();
+  HandleScope scope(isolate);
+
+  MaybeLocal<String> hint_str = String::NewFromUtf8(
+    isolate,
+    identity,
+    v8::NewStringType::kNormal,
+    strlen(identity));
+
+  if (hint_str.IsEmpty()) {
+    return 0;
+  }
+
+  Local<Value> argv[] = {
+    hint_str.ToLocalChecked(),
+    Integer::NewFromUnsigned(isolate, max_psk_len),
+    Integer::NewFromUnsigned(isolate, 0)
+  };
+
+  Local<Value> value = sc->object()->Get(env->onpskexchange_string());
+
+  MaybeLocal<Value> maybe_ret;
+  if (value->IsFunction()) {
+    Local<Function> func = Local<Function>::Cast(value);
+    maybe_ret = func->Call(env->context(), sc->object(), arraysize(argv), argv);
+  } else {
+    return 0;
+  }
+
+  if (maybe_ret.IsEmpty()) {
+    return 0;
+  }
+
+  Local<Value> ret = maybe_ret.ToLocalChecked();
+
+  // The result is expected to be an object. If it isn't, then return 0,
+  // indicating the identity wasn't found.
+  if (!ret->IsObject()) {
+    return 0;
+  }
+  Local<Object> obj = ret.As<Object>();
+
+  Local<Value> psk_val = obj->Get(env->context(),
+                                  env->psk_string()).ToLocalChecked();
+  CHECK(psk_val->IsString());
+  Local<String> psk_str = psk_val.As<String>();
+
+  Local<String> enc = String::Empty(isolate);
+  StringBytes::InlineDecoder decoder;
+  if (!decoder.Decode(env, psk_str, enc, HEX))
+    return 0;
+
+  char* psk_buf = decoder.out();
+  size_t psk_len = decoder.size();
+
+  CHECK_LE(psk_len, max_psk_len);
+  memcpy(psk, psk_buf, psk_len);
+
+  return psk_len;
+}
+
+unsigned int SecureContext::PskClientCallback(SSL *ssl,
+                                              const char *hint,
+                                              char *identity,
+                                              unsigned int max_identity_len,
+                                              unsigned char *psk,
+                                              unsigned int max_psk_len) {
+  SecureContext* sc = static_cast<SecureContext*>(
+      SSL_CTX_get_app_data(ssl->ctx));
+
+  Environment* env = sc->env();
+  Isolate* isolate = env->isolate();
+  HandleScope scope(isolate);
+
+  Local<Value> argv[] = {
+    Null(isolate),
```

Comment:
May be helpful to add a comment here reminding folks that this is a error-back style callback, which explains why the first argument is passed as a `Null`

Reply:
So that's actually the object passed to the `pskCallback` function, where the first argument (initially defined as `Null`) is the hint provided by OpenSSL. The hint is an optional field, which is why we're not erroring when it's not provided.